pkg/reconciler/apis/apiexport/apiexport_controller.go

```diff
@@ -0,0 +1,465 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apiexport
+
+import (
+	"context"
+	cryptorand "crypto/rand"
+	"crypto/rsa"
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	jsonpatch "github.com/evanphx/json-patch"
+	"github.com/kcp-dev/apimachinery/pkg/logicalcluster"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/kubernetes"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clusters"
+	"k8s.io/client-go/util/keyutil"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	kcpclient "github.com/kcp-dev/kcp/pkg/client/clientset/versioned"
+	apisinformers "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/apis/v1alpha1"
+	apislisters "github.com/kcp-dev/kcp/pkg/client/listers/apis/v1alpha1"
+	conditionsv1alpha1 "github.com/kcp-dev/kcp/third_party/conditions/apis/conditions/v1alpha1"
+	"github.com/kcp-dev/kcp/third_party/conditions/util/conditions"
+)
+
+const (
+	controllerName = "kcp-apiexport"
+
+	annotationKeyAPIExportCluster = "apis.kcp.dev/apiexport-cluster"
+	annotationKeyAPIExportName    = "apis.kcp.dev/apiexport-name"
+
+	identityKey = "key"
+)
+
+// NewController returns a new controller for APIExports.
+func NewController(
+	kcpClusterClient kcpclient.ClusterInterface,
+	apiExportInformer apisinformers.APIExportInformer,
+	kubeClusterClient kubernetes.ClusterInterface,
+	namespaceInformer coreinformers.NamespaceInformer,
+	secretInformer coreinformers.SecretInformer,
+) (*controller, error) {
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), controllerName)
+
+	c := &controller{
+		queue:             queue,
+		kcpClusterClient:  kcpClusterClient,
+		apiExportLister:   apiExportInformer.Lister(),
+		kubeClusterClient: kubeClusterClient,
+		getNamespace: func(clusterName logicalcluster.LogicalCluster, name string) (*corev1.Namespace, error) {
+			return namespaceInformer.Lister().Get(clusters.ToClusterAwareKey(clusterName, name))
+		},
+		createNamespace: func(ctx context.Context, clusterName logicalcluster.LogicalCluster, ns *corev1.Namespace) error {
+			_, err := kubeClusterClient.Cluster(clusterName).CoreV1().Namespaces().Create(ctx, ns, metav1.CreateOptions{})
+			return err
+		},
+		secretLister:    secretInformer.Lister(),
+		secretNamespace: "default",
```

Comment:
this might be a security problem. Would rather prefer a namespace like kube-system where users have no secret-read access.

Reply:
Use case: user has create-apiexport permission but not read-secrets?

Reply:
Changed to kcp-system